sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/typedaggregators.scala

```diff
@@ -99,3 +94,91 @@ class TypedAverage[IN](val f: IN => Double) extends Aggregator[IN, (Double, Long
     toColumn.asInstanceOf[TypedColumn[IN, java.lang.Double]]
   }
 }
+
+class TypedMinDouble[IN](val f: IN => Double) extends Aggregator[IN, Double, Double] {
+  override def zero: Double = Double.PositiveInfinity
+  override def reduce(b: Double, a: IN): Double = math.min(b, f(a))
+  override def merge(b1: Double, b2: Double): Double = math.min(b1, b2)
+  override def finish(reduction: Double): Double = {
+    if (Double.PositiveInfinity == reduction) {
```

Comment:
If the input is one `Double.PositiveInfinity`, we will get return `Double.NegativeInfinity` as result?

Reply:
That's correct. That was part of the discussion above. We used to init it with null, so that we could then distinguish between these cases. As you can read above, that initial proposal was tossed as it didnt meet ANSI standards.
Another option I just realised would be to initialize it with Double.NaN, and then use that as a flag to distinguish between infinity and the initial value. Then again, that would not be supported for Longs as we cannot assign a NaN.

Reply:
Since it's hard to tell if it's empty input or inputs of all `Double.PositiveInfinity`, my new proposal
```
class MutableLong(var value: Long) extend Serializable

class TypedMinLong[IN](val f: IN => Long) extends Aggregator[IN, MutableLong, java.lang.Long] {
  override def zero: MutableLong = null
  override def reduce(b: MutableLong, a: IN): MutableLong = {
    if (b == null) {
      new MutableLong(f(a))
    } else {
      b.value = math.max(b.value, f(a))
      b
    }
  }
  override def finish(reduction: MutableLong): java.lang.Long = {
    if (reduction == null) {
      null
    } else {
      reduction.value
    }
  }
}
```

Reply:
Doesn't that boil down to what was there previously? https://github.com/apache/spark/pull/18113/commits/51783b55197cea6c130722838ec97ad6df5c92be

```
  override def zero: java.lang.Double = null
  override def reduce(b: java.lang.Double, a: IN): java.lang.Double =
    if (b == null) f(a) else math.max(b, f(a))

  override def merge(b1: java.lang.Double, b2: java.lang.Double): java.lang.Double = {
    if (b1 == null) {
      b2
    } else if (b2 == null) {
      b1
    } else {
      math.max(b1, b2)
    }
  }
  override def finish(reduction: java.lang.Double): java.lang.Double = reduction
```

Here we just return null in case its an empty set or if we have the edge case you just mentioned. You rejected it because you were afraid of boxing performance on the 8th of June.

Reply:
The thing is that, using `java.lang.Long` as buffer type will introduce boxing and unboxing for every computation, while `MutableLong` doesn't.

Reply:
Ok makes sense. What about the return ```finish``` return type? Leaving that as a java type would cause the ```this``` and ```toColumnJava``` to be flipped, creating a  ```toColumnScala``` instead.
What about:
```
override def finish(reduction: java.lang.Double): Double = reduction
```
As its on the finish, it shouldn't cause much performance overhead as its not execution many times. It would also reduce complexity a bit.

Reply:
I'm ok with
```
  override def finish(reduction: MutableLong): Long = {
    if (reduction == null) {
      Long.MinValue
    } else {
      reduction.value
    }
  }
```

Reply:
This is more scala-friendly, and the semantic looks reasonable as we discussed before: min of empty input is Long/Double.MinValue

Reply:
Actually we have 3 choices for empty input here:
1. return null and the return type is java.lang.Long
2. return Long.MinValue and return type is primitive long
3. throw exception and return type is primitive long

cc @gatorsmile @HyukjinKwon @srowen

Reply:
Ah sorry yes, was only talking about the output type. 
I don't have time right now but Ill do it over the weekend and mark you when I am done.

Reply:
Which of those 3 will we decide on then? None of them is ideal unfortunately.

cc @cloud-fan @gatorsmile @HyukjinKwon @srowen

Reply:
After some more thoughts, options 3 is not reasonable as throwing exception is not a good idea in big data, especially in the last stage of a long-running job.

option 2 is weird as it doesn't follow either java/scala or SQL.

Let's go with option 1.

Reply:
Oh, I don't know why I missed this cc. Will check though related stuff and be back soon.

Reply:
+1 for option 1. It sounds the best among three to me too.